core/src/host.rs

```diff
@@ -0,0 +1,673 @@
+    ///
+    /// If there are tokens attached to the current instance, the tokens will be attached
+    /// to the next contract call.
+    fn with_tokens(&self, tokens: U512) -> Self;
+    /// Returns the address of the contract.
+    fn address(&self) -> &Address;
+    /// Returns the host environment.
+    fn env(&self) -> &HostEnv;
+    /// Returns the n-th event emitted by the contract.
+    ///
+    /// If the event is not found or the type does not match, returns `EventError::EventNotFound`.
+    fn get_event<T: 'static>(&self, index: i32) -> Result<T, EventError>
+    where
+        T: FromBytes + EventInstance;
+    /// Returns a detailed information about the last call of the contract.
+    fn last_call(&self) -> ContractCallResult;
+}
+
+/// Trait for loading a contract from the host environment.
+///
+/// Similar to [Deployer], but does not deploy a new contract, but loads an existing one.
+pub trait HostRefLoader {
+    /// Loads an existing contract from the host environment.
+    fn load(env: &HostEnv, address: Address) -> Self;
+}
+
+/// A type which can provide an [EntryPointsCaller].
+pub trait EntryPointsCallerProvider {
+    /// Returns an [EntryPointsCaller] for the given host environment.
+    fn entry_points_caller(env: &HostEnv) -> EntryPointsCaller;
+}
+
+/// A type which can deploy a contract.
+///
+/// Before any interaction with the contract, it must be deployed, either
+/// on a virtual machine or on a real blockchain.
+///
+/// The `Deployer` trait provides a simple way to deploy a contract.
+pub trait Deployer {
+    /// Deploys a contract with given init args.
+    ///
+    /// If the init args are provided, the contract is deployed and initialized
+    /// by calling the constructor. If the init args are not provided, the contract
+    /// is deployed without initialization.
+    ///
+    /// Returns a host reference to the deployed contract.
+    fn deploy<T: InitArgs>(env: &HostEnv, init_args: T) -> Self;
+}
+
+/// A type which can be used as initialization arguments for a contract.
+#[cfg_attr(test, mockall::automock)]
+pub trait InitArgs {
+    /// Validates the args are used to initialized the right contact.
+    ///
+    /// If the `expected_ident` does not match the contract ident, the method returns `false`.
+    fn validate(expected_ident: &str) -> bool;
+    /// Converts the init args into a [RuntimeArgs] instance.
+    fn into_runtime_args(self) -> Option<RuntimeArgs>;
+}
+
+/// Default implementation of [InitArgs]. Should be used when the contract
+/// has the constructor but does not require any initialization arguments.
+pub struct EmptyInitArgs;
+
+impl InitArgs for EmptyInitArgs {
+    fn validate(_expected_ident: &str) -> bool {
+        true
+    }
+
+    fn into_runtime_args(self) -> Option<RuntimeArgs> {
+        Some(RuntimeArgs::new())
+    }
+}
+
+/// An implementation of [InitArgs]. Should be used when the contract
+/// does not have the constructor.
+pub struct NoneInitArgs;
```

Comment:
```suggestion
pub struct NoInit;
```
I think `NoInit` is better as it suggests that there is no init function.